proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+```c# 
+// Existing Span<T> definition
+public readonly ref struct Span<T>
+{
+    public Span(T[]? array) { ... }
+
+    public Span(T[]? array, int index, int length) { ... }
+
+    public static implicit operator Span<T>(T[]? array) { ... }
+}
+```
+
+If the `Span<T>` type was being designed from scratch in the presence of `where T : ref struct` it's possible different choices would've been made here. Possibly leveraging `static` factories and pushing implicit conversions into the compiler. Unfortunately this is not being designed from scratch and in order to meet the desire to have `Span<Span<T>>` these existing APIs need to be rationalized.
+
+The compiler will introduce the type `System.Runtime.CompilerServices.SupportsOnlyNonByRefLikeAttribute` that can be applied to methods and operators. When applied to a member it has the effect of removing `where T : ref struct` from the constraints on all type parameters on the *immediate* containing type. All other constraints remain in place.
+
+This allows us to rationalize our `Span<T>` definition as well as allowing third party customers rationalize their existing APIs. 
+
+```c#
+// Existing Span<T> definition
+public readonly ref struct Span<T>
+{
+    [SupportsOnlyNonByRefLikeAttributes]
+    public Span(T[]? array) { ... }
+
+    [SupportsOnlyNonByRefLikeAttributes]
+    public Span(T[]? array, int index, int length) { ... }
+
+    [SupportsOnlyNonByRefLikeAttributes]
+    public static implicit operator Span<T>(T[]? array) { ... }
+}
+```
+
+This does not impact the binding rules for invocations of such members. These symbols will continue to be fully bindable. In the case they are bound when any of the type parameters on the containing type are potentially a `ref struct` then use of a member annotated with `[SupportsOnlyNonByRefLike]` will be considered an error.
+
+```c#
+void M<T, U>()
+    where T : ref struct
+{
+    // Error: cannot access member as `T` is potentially a `ref struct`
+    new Span<T>(null);
+
+    // Okay
+    new Span<U>(null);
+}
+```
+
+The runtime will recognize this attribute and ensure that any use of these members at runtime with a type that is a `ref struct` will result in a runtime exception.
+
+Note: the [alternative](#special-case) here is simply special casing `Span<T>` as array is special cased today.
+
+### ref struct interfaces
+<a name="interfaces"></a>
+
+The language will allow `ref struct` to implement interfaces. The rules for implementation will be the same as for normal `struct`. This includes allowing the implementations to be implicit or explicit.
+
+**TODO: Need better example than animals but feeling lazy**
+
+```c#
+interface IAnimal
+{
+    void Speak(); 
+}
+
+ref struct Lab : IAnimal
+{
+    public void Speak() => Console.WriteLine("FEED ME");
```

Comment:
Was Penny that hungry when you were writing this?

Reply:
Always